app/questionnaire/questionnaire_manager.py

```diff
@@ -328,62 +284,71 @@ def process_incoming_answers(self, location, post_data, replay=False):
 
             # now return the location
             current_location = self.get_current_location()
             logger.debug("Returning location %s", current_location)
-            return current_location
+            return is_valid
         else:
             raise InvalidLocationException()
 
-    def get_rendering_context(self, location):
-
-        # apply any conditional display rules
-        self._conditional_display(self.get_state(location).state)
+    def get_rendering_context(self, location, is_valid=True):
 
         # look up the preprocessor and then build the view data
         preprocessor = TemplateRegistry.get_template_preprocessor(location)
-
-        if location == 'summary':
-            # the summary is the special case when we need the start of the linked list
-            node = self._first
-            # and we also need to plumb the entire schema
-            while node.next:
-                self._plumbing_preprocessing(node)
-                node = node.next
-
-            # reset pointer back to the first node for the preprocessor
-            node = self._first
-
-        else:
-            # unlike the rest where we need the current node in the list
-            node = self.get_state(location)
-            # and only need to plumb the single page
-            self._plumbing_preprocessing(node)
-        return preprocessor.build_view_data(node, self._schema)
-
-    def _plumbing_preprocessing(self, node):
-        '''
-        Run the current state through the plumbing preprocessor
-        :return:
-        '''
+        node = self._current
+        state_items = []
+        if is_valid:
+            if location == 'summary':
+                # the summary is the special case when we need the start of the linked list
+                node = self._first
+                # and we also need to plumb the entire schema
+                while node.next:
+                    self.build_state(node, node.answers)
+                    if self.schema_item:
+                        self._plumbing_preprocessing(self.state)
+                        self._conditional_display(self.state)
+                        state_items.append(self.state)
+                    node = node.next
+                # reset pointer back to the first node for the preprocessor
+                node = self._first
+            else:
+                self.build_state(node, node.answers)
+
+        if self.schema_item:
+            self._plumbing_preprocessing(self.state)
+            self._conditional_display(self.state)
+        state_items.append(self.state)
+
+        return preprocessor.build_view_data(node, self._schema, state_items)
+
+    def build_state(self, node, answers):
+        # Build the state from the linked list and the answers
+        self.state = None
+        self.schema_item = None
+        if self._schema.item_exists(node.item_id):
+            self.schema_item = self._schema.get_item_by_id(node.item_id)
```

Comment:
I don't think we need to set schema_item on self. Once we've created the state, isn't it accessible via `state.schema_item`?